Refactored the term suggestion builder for the query refactoring effort

abeyad opened the pull request:
Added the term suggestion builder's serialization/deserialization and
equals/hashCode methods.

cbuescher commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -69,17 +69,9 @@
-    protected StreamInput(NamedWriteableRegistry namedWriteableRegistry) {
-        this.namedWriteableRegistry = namedWriteableRegistry;
-    }
+    protected StreamInput() { }
```
This looks like a leftover indeed an I think it can be removed, but maybe @javanna can take a second look?

javanna replied:
what is unused is unused :) I have no idea how it became unused though.

cbuescher replied:
I think NamedWriteableAwareStreamInput was introduced just a bit later. Only my guess.

cbuescher commented on core/src/main/java/org/elasticsearch/common/util/NamedEnum.java:
```diff
@@ -0,0 +1,64 @@
+     * @return The enum value matching the given name
+     * @throws IllegalArgumentException if the input is invalid or an enum value with the given name could not be found
+     */
+    static <E extends Enum<E>> E fromString(final Class<E> cls, final String name) {
```
+1 on having a generalized fromString() method for Enums and a common interface for this. I don't quiet recall why we serialize the enum ordinals in lots of other cases (probably size considerations?) but I'm wondering if we should go over other enums we used in the refactoring so far and maybe change them? That should be a broader discussion though.

cbuescher commented on core/src/main/java/org/elasticsearch/common/util/NamedEnum.java:
```diff
@@ -0,0 +1,64 @@
+            return null;
+        }
+        checkNotNull(cls, "class");
+        if (!NamedEnum.class.isAssignableFrom(cls)) {
```
Just curious if you have any experience about performance with this checks, and the `allOf` call below. Maybe negligable, but since we will call this in serialization if we use the enum names there, it might be called a lot.

cbuescher commented on core/src/main/java/org/elasticsearch/common/util/Preconditions.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Checks that the argument is not null; otherwise, throws an <code>IllegalArgumentException</code>
+     */
+    public static void checkNotNull(final Object obj, @Nullable final String argName) {
```
This can probably be replaced by requireNonNull() and a Supplier? Just trying to reduce the need for too many helper classes.

cbuescher commented on core/src/main/java/org/elasticsearch/search/suggest/SuggestBuilder.java:
```diff
@@ -19,31 +19,37 @@
+ * to the terms in provided text. These suggestions are based on several options described in this class.
  */
-public class SuggestBuilder extends ToXContentToBytes {
+public class SuggestBuilder extends ToXContentToBytes implements NamedWriteable<SuggestBuilder> {
```
I think we don't need to make the global SuggestBuilder implement NamedWritable, simply Writable should be enough. This is the only implementation (and will likely stay so) so we don't need to differentiate it from others on the stream.

cbuescher commented on core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java:
```diff
@@ -296,4 +297,72 @@ public final int hashCode() {
+
+
+    /** An enum representing the valid suggest modes. */
+    public enum SuggestMode implements NamedEnum {
```
Its great to have these as enums, especially with all the added javadoc. If they are specific to the TermsSuggestionBuilder (haven't checked), I would move them there though.

cbuescher commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -117,33 +150,59 @@ public TermSuggestionBuilder sort(String sort) {
      */
-    public TermSuggestionBuilder maxEdits(Integer maxEdits) {
+    public TermSuggestionBuilder maxEdits(int maxEdits) {
+        checkArgument(maxEdits > 0 && maxEdits < 3, "maxEdits");
```
I like the succinctness of this checks, wondering if java 8 offers anything to avoid having to introduce the utility class for this though.

cbuescher commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -306,6 +306,11 @@ public static double randomDoubleBetween(double start, double end, boolean lower
     }
 
+    /** A random float between 0.0 and 1.0 (inclusive). */
+    public static float randomFloatRatio() {
```
Whats the difference of this to randomFloat()? To me it looks like that would do the trick, but might be missing something.

abeyad replied:
@cbuescher The issue with randomFloat() is that it could return any float number, positive or negative and in any range.  I wanted something that gives a float between 0.0 and 1.0, and I figured many of our API values are ranges between 0.0 and 1.0.  Do you think it warrants its own call  in ESTestCase or should I just put the same functionality in SuggestSearchTests?

cbuescher replied:
@abeyad randomFloat() should do just that. The method name suggests something more generic but if you check it delegates to Random.nextFloat() which has some docs. I also just re-checked, just so make sure.

abeyad replied:
@cbuescher you're right, I didn't realize that!  Thanks for catching it :)

cbuescher commented on core/src/test/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilderTests.java:
```diff
@@ -0,0 +1,118 @@
+            case 0: return SuggestMode.MISSING;
+            case 1: return SuggestMode.POPULAR;
+            case 2: return SuggestMode.ALWAYS;
+            default: return SuggestMode.MISSING; // should never happen
```
Either leave it or throw an error here, so someone takes a look at the test.

cbuescher commented:
Looks great already, I left some comments and maybe @areek wants to have another look too.

cbuescher commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -187,10 +273,20 @@ public TermSuggestionBuilder minWordLength(int minWordLength) {
      * value higher than 1 is specified then the number cannot be fractional.
      */
     public TermSuggestionBuilder minDocFreq(float minDocFreq) {
+        checkArgument(minDocFreq > 0.0f && (minDocFreq <= 1.0f || minDocFreq == Math.floor(minDocFreq)), "minDocFreq");
```
A value if 0 seems to be okay here. At least there are some tests in SuggestSearchTests (unfortunately in the lang-mustache module) that use that and now fail with an exception.

abeyad commented:
@cbuescher I made the changes we discussed.

abeyad commented on core/src/main/java/org/elasticsearch/common/util/NamedEnum.java:
```diff
@@ -0,0 +1,57 @@
+     * @param name The name of the enum within the given type
+     * @return The enum value matching the given name
+     * @throws IllegalArgumentException if the input is invalid or an enum value with the given name could not be found
+     */
```
@s1monw Wanted to get your input on the method below.  Its meant to be a convenience method for taking an input string (e.g. a JSON parameter) and converting it to an enum, in the situation where the enum element does not have the same name as the input string (and hence, can't use Enum.valueOf).  A common case is where we name the enum element in all upper case but the string value is in lower case.

s1monw replied:
I like the trick here but I think it's too much magic after all. Can we stick to switch / case and ordinals - it's something everybody understands without reading up on it? WDYT

cbuescher replied:
Plus I just realized that sometimes we have more than one variant of a String in the query DSL (e.g. camel casing etcll) that needs to be mapped to one enum. For example, the StringDistance currently allows both `"damerau_levenshtein"` and  `"damerauLevenshtein"`. This might be something to get rid of, but in the meantime we need to support it I think.

abeyad replied:
@s1monw @cbuescher Good points, especially with the multiple variants for any given enum value that Christoph mentioned, this interface provides no added benefit then.  I'll take it out.

cbuescher commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -233,25 +347,128 @@ public String getWriteableName() {
+        out.writeOptionalVInt(suggestMode != null ? suggestMode.ordinal() : null);
+        out.writeOptionalFloat(accuracy);
+        out.writeOptionalVInt(sort != null ? sort.ordinal() : null);
+        out.writeOptionalVInt(stringDistance != null ? stringDistance.ordinal() : null);
```
Can you put the read/write logic for the enums into the respective class? That way it's easier to see how these enums are sent over the wire.

abeyad replied:
Yes, will do.

cbuescher commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -187,10 +286,25 @@ public TermSuggestionBuilder minWordLength(int minWordLength) {
      * value higher than 1 is specified then the number cannot be fractional.
      */
     public TermSuggestionBuilder minDocFreq(float minDocFreq) {
+        if (minDocFreq < 0.0f) {
```
Can you put a few tests for illegal values (one for each setter that has arg-checks is enough I think) into TermSuggestionBuilderTests?

abeyad replied:
Good catch, will add those tests.

cbuescher commented:
@abeyad thanks, looks very good now. I left two minor comments, also lets wait on input on the NamedEnum, then I think this is ready for merging.

abeyad commented:
@cbuescher @colings86 @areek I updated the PR with the latest commits that address the code review comments.  One thing to note is that I also added a `AbstractWriteableEnumTestCase` that reduces the amount of boilerplate code for writing tests for enums that implement `Writeable`.

colings86 commented on core/src/main/java/org/elasticsearch/search/suggest/SuggestBuilder.java:
```diff
@@ -19,31 +19,35 @@
     }
 
-    public SuggestBuilder(String name) {
+    public SuggestBuilder(@Nullable String name) {
```
This probably shouldn't be nullable. We have the zero-arg constructor above for when a no name is needed so maybe we should throw an exception if the name is null here so the user is alerted to the fact that the variable they are passing in is null?

cbuescher replied:
Good point, but my personal opinion here is that if the value is optional we should allow null. It would be different if this can overwrite a default settings, but thats not the case here. I'm not a big fan of the annotation though.

abeyad replied:
I figured the annotation would declare the intent clearly.  I think the bigger issue is deciding if `name` can be null or not and then build a consistent API for the builder that allows to set the value consistently (i.e. everywhere the property can either be null or everywhere it is prohibited from being null).  Just my two cents.

colings86 commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -65,12 +72,20 @@ public TermSuggestionBuilder(String name) {
      * </ol>
      */
-    public TermSuggestionBuilder suggestMode(String suggestMode) {
+    public TermSuggestionBuilder suggestMode(@Nullable SuggestMode suggestMode) {
```
I don't think we should allow null's to be passed into Builders as this can easily lead to NPEs. If this settings are not required then the user should not set them rather than setting them to null.

cbuescher replied:
I'm on the fence with this. In cases where the value is optional I don't think we need to make it a hard rule to check for null. We did it in several place in the queries, but there are also many setters without explicit null-checks, wdyt?

colings86 replied:
I commented above on defaults (https://github.com/elastic/elasticsearch/pull/16278/files#r51552740) and I think where we have defaults we should make sure they cannot be nulled as this will create hard to debug errors for the user. IMO we should stick to the mantra of not allowing an invalid builder object to be built in the Java API so that the user can easily see where they have made errors in their own code.

abeyad replied:
This is a very good point about not allowing the builder to be in an invalid state.

abeyad replied:
The `DirectSpellcheckerSettings` contains some hard coded values which we can put into constants that both the builder and the settings use.  Is there a convention for where to put those constants - in builders or in these "settings" objects or elsewhere?

colings86 replied:
Wherever makes the most sense really. In this case I would put the default constants in `DirectSpellcheckerSettings` I think

colings86 commented on core/src/test/java/org/elasticsearch/common/io/stream/AbstractWriteableEnumTestCase.java:
```diff
@@ -0,0 +1,64 @@
+        doWriteTo();
+    }
+
+    protected abstract void doReadFrom() throws IOException;
```
Why do we need a concrete and an abstract method for each test? I think we should just make `testReadFrom` and `testWriteTo` abstract.

abeyad replied:
You're right, i started down the road of having other code in `testReadFrom()` and `testWriteTo()`, then calling into the abstract method, but that's no longer there now.  I'll clean this up.

colings86 commented on core/src/test/java/org/elasticsearch/common/io/stream/AbstractWriteableEnumTestCase.java:
```diff
@@ -0,0 +1,64 @@
+    protected abstract void doWriteTo() throws IOException;
+
+    // a convenience method for testing the write of a writeable enum
+    protected void writeToStream(final Writeable writeableEnum, final int ordinal) throws IOException {
```
Can we rename this to `assertWriteToStream` and the below method to `assertReadFromStream` so that it's clear that they both serialize and check the enum instead of just serialising it?

abeyad replied:
Will do

colings86 commented on core/src/test/java/org/elasticsearch/common/io/stream/AbstractWriteableEnumTestCase.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Abstract class offering base functionality for testing @{link Writeable} enums.
+ */
+public abstract class AbstractWriteableEnumTestCase extends ESTestCase {
```
We should also have tests for testing the ordinal value of each enum value (`Enum.ordinal`) and for checking that reading a random invalid ordinal throws an exception. See ShapeRelationTests for examples of what I mean

abeyad replied:
Those tests are in the concrete classes, you mean putting an abstract method in there to force the concrete test to implement those test functions?

colings86 replied:
yes

colings86 commented on core/src/test/java/org/elasticsearch/common/io/stream/AbstractWriteableEnumTestCase.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    protected <T extends Writeable<T>> void readFromStream(final int ordinal, final Writeable<T> expected) throws IOException {
```
I think this method and the method above could be static?

abeyad replied:
Yep, i'll make the change.

colings86 commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -31,12 +37,13 @@
+    private StringDistanceImpl stringDistance;
     private Integer maxEdits;
     private Integer maxInspections;
     private Float maxTermFreq;
```
These should all have sensible defaults rather than being null, as we have done with all the other builders. This also means we can remove @Nullable from all of the methods and add checks in there that throw an exception if null is passed in to make it safer. The defaults we currently use can be found in DirectSpellcheckerSettings.

abeyad commented:
@cbuescher @colings86 @areek Pushed changes based on discussion earlier.

abeyad commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -233,25 +358,185 @@ public String getWriteableName() {
+        }
+
+        // TODO: move this to SuggestUtils when the builder becomes the intermediate representation
+        static SuggestMode fromUnderlying(final org.apache.lucene.search.spell.SuggestMode mode) {
```
@cbuescher @colings86 @areek Just FYI, I added this in here so the builder's defaults are appropriately based on the `DefaultSpellcheckerSettings` defaults, but I expect this to move to `SuggestUtils` when we refactor the JSON parsing, etc.  I also believe these enums will go through some refactoring, for example there is no reason to have both the SortBy enum below and `org.elasticsearch.search.suggest.Suggest.Sort` .. they should be consolidated into a single enum from what I can tell, but that will be done in future PRs.

areek replied:
@abeyad I don't understand why we need `fromUnderlying` at all on the writable enums, doesn't it makes sense just to hardcode the corresponding `DefaultSpellcheckerSettings` defaults? maybe a comment would be enough in `DefaultSpellcheckerSettings` to note that the defaults should be in sync with `TermSuggestionBuilder`, as these defaults don't change too often if at all? 

> I also believe these enums will go through some refactoring, for example there is no reason to have both the SortBy enum below and org.elasticsearch.search.suggest.Suggest.Sort .. they should be consolidated into a single enum from what I can tell, but that will be done in future PRs.

++

cbuescher replied:
Also, we will need the oposite conversion from the ES enums (e.g. TermSuggestionBuilder.SuggestMode) to the Lucene enums (org.apache.lucene.search.spell.SuggestMode) used in the DirectSpellcheckerSettings later anyway, so rather than having `fromUnderlying` better turn the direction around to each enum knows how to produce the corresponding low-level enum.

abeyad replied:
@areek @cbuescher good points, I put them in there to force us to revisit the default retrieval when we refactor, but I agree that its overkill and we can just put a comment in the code to that effect.  I'll remove this.

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -65,12 +89,20 @@ public TermSuggestionBuilder(String name) {
      */
-    public TermSuggestionBuilder suggestMode(String suggestMode) {
+    public TermSuggestionBuilder suggestMode(SuggestMode suggestMode) {
+        Objects.requireNonNull(suggestMode, "suggest_mode cannot be null");
```
nit: IMO `suggestMode must not be null` sounds more explicit.

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -78,31 +110,49 @@ public TermSuggestionBuilder suggestMode(String suggestMode) {
-    public TermSuggestionBuilder setAccuracy(float accuracy) {
+    public TermSuggestionBuilder accuracy(float accuracy) {
+        if (accuracy < 0.0f || accuracy > 1.0f) {
+            throw new IllegalArgumentException("Accuracy must be between 0 and 1");
```
nit: `accuracy` instead of `Accuracy`

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -78,31 +110,49 @@ public TermSuggestionBuilder suggestMode(String suggestMode) {
      */
-    public TermSuggestionBuilder sort(String sort) {
+    public TermSuggestionBuilder sort(SortBy sort) {
+        Objects.requireNonNull(sort, "sort cannot be null");
```
nit: IMO `sort must not be null`

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -117,33 +167,61 @@ public TermSuggestionBuilder sort(String sort) {
      */
-    public TermSuggestionBuilder stringDistance(String stringDistance) {
+    public TermSuggestionBuilder stringDistance(StringDistanceImpl stringDistance) {
+        Objects.requireNonNull(stringDistance, "string_distance cannot be null");
```
nit: `stringDistance` must not be null`

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -117,33 +167,61 @@ public TermSuggestionBuilder sort(String sort) {
-    public TermSuggestionBuilder maxEdits(Integer maxEdits) {
+    public TermSuggestionBuilder maxEdits(int maxEdits) {
+        if (maxEdits < 1 || maxEdits > 2) {
+            throw new IllegalArgumentException("max_edits must be between 1 and 2");
```
`maxEdits` instead of `max_edits`

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -117,33 +167,61 @@ public TermSuggestionBuilder sort(String sort) {
-    public TermSuggestionBuilder maxInspections(Integer maxInspections) {
+    public TermSuggestionBuilder maxInspections(int maxInspections) {
+        if (maxInspections < 0) {
+            throw new IllegalArgumentException("max_inspections must be a positive value");
```
nit: `maxInspections must be positive`

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -155,31 +233,64 @@ public TermSuggestionBuilder maxInspections(Integer maxInspections) {
      */
     public TermSuggestionBuilder prefixLength(int prefixLength) {
+        if (prefixLength < 0) {
+            throw new IllegalArgumentException("prefix_length must be a positive value");
```
`prefixLength` must be positive

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -155,31 +233,64 @@ public TermSuggestionBuilder maxInspections(Integer maxInspections) {
      */
     public TermSuggestionBuilder minWordLength(int minWordLength) {
+        if (minWordLength < 1) {
+            throw new IllegalArgumentException("min_word_length must be greater than or equal to 1");
```
minWordLength must be greater or equal to 1

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -187,10 +298,24 @@ public TermSuggestionBuilder minWordLength(int minWordLength) {
      */
     public TermSuggestionBuilder minDocFreq(float minDocFreq) {
+        if (minDocFreq < 0.0f) {
+            throw new IllegalArgumentException("min_doc_freq cannot be a negative value");
```
minDocFreq must be positive

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -233,25 +358,185 @@ public String getWriteableName() {
+        }
+    }
+
+    /** An enum represting the valid string edit distance algorithms for determining suggestions. */
```
nit: typo `representing`

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -155,31 +233,64 @@ public TermSuggestionBuilder maxInspections(Integer maxInspections) {
      */
     public TermSuggestionBuilder maxTermFreq(float maxTermFreq) {
+        if (maxTermFreq < 0.0f) {
+            throw new IllegalArgumentException("max_term_freq cannot be less than 0");
```
nit: maxTermFreq must be positive

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -155,31 +233,64 @@ public TermSuggestionBuilder maxInspections(Integer maxInspections) {
+            throw new IllegalArgumentException("max_term_freq cannot be less than 0");
+        }
+        if (maxTermFreq > 1.0f && maxTermFreq != Math.floor(maxTermFreq)) {
+            throw new IllegalArgumentException("If max_term_freq is greater than 1, then it cannot be a fractional");
```
nit: `if maxTermFreq is greater than 1, it must not be a fraction`

areek commented on core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java:
```diff
@@ -187,10 +298,24 @@ public TermSuggestionBuilder minWordLength(int minWordLength) {
+            throw new IllegalArgumentException("min_doc_freq cannot be a negative value");
+        }
+        if (minDocFreq > 1.0f && minDocFreq != Math.floor(minDocFreq)) {
+            throw new IllegalArgumentException("If min_doc_freq is greater than 1, it cannot be a fractional");
```
nit: `if minDocFreq is greater than 1, then it must not be a fraction`

areek commented:
thanks @abeyad this looks great! I left some minor comments.

abeyad commented:
@areek @cbuescher @colings86 I removed `fromUnderlying` and fixed the error messages per Areek's comments.  If its good otherwise, should we merge?

areek commented:
Thanks @abeyad LGTM!
